src/patcher/java/net/minecraftforge/gradle/patcher/PatcherPlugin.java

```diff
@@ -139,156 +150,143 @@ public void apply(@Nonnull Project project) {
             });
         });
+
         new BaseRepo.Builder()
-            .add(MCPRepo.create(project))
-            .add(MinecraftRepo.create(project))
-            .attach(project);
+                .add(MCPRepo.create(project))
+                .add(MinecraftRepo.create(project))
+                .attach(project);
         project.getRepositories().maven(e -> {
             e.setUrl(Utils.MOJANG_MAVEN);
             e.metadataSources(MetadataSources::artifact);
         });
 
-        hideLicense.configure(task -> {
-            task.doLast(_task -> {
-                MojangLicenseHelper.hide(project, extension.getMappingChannel(), extension.getMappingVersion());
-            });
-        });
+        hideLicense.configure(task -> task.doLast(_task ->
+                MojangLicenseHelper.hide(project, extension.getMappingChannel().get(), extension.getMappingVersion().get())));
 
-        showLicense.configure(task -> {
-            task.doLast(_task -> {
-                MojangLicenseHelper.show(project, extension.getMappingChannel(), extension.getMappingVersion());
-            });
-        });
+        showLicense.configure(task -> task.doLast(_task ->
+                MojangLicenseHelper.show(project, extension.getMappingChannel().get(), extension.getMappingVersion().get())));
+
+        release.configure(task -> task.dependsOn(sourcesJar, universalJar, userdevJar));
+
+        dlMappingsConfig.configure(task -> task.getMappings().convention(extension.getMappings()));
 
-        release.configure(task -> {
-            task.dependsOn(sourcesJar, universalJar, userdevJar);
-        });
-        dlMappingsConfig.configure(task -> {
-            task.setMappings(extension.getMappings());
-        });
         extractNatives.configure(task -> {
-            task.dependsOn(dlMCMetaConfig.get());
-            task.setMeta(dlMCMetaConfig.get().getOutput());
-            task.setOutput(natives_folder);
-        });
-        downloadAssets.configure(task -> {
-            task.dependsOn(dlMCMetaConfig.get());
-            task.setMeta(dlMCMetaConfig.get().getOutput());
+            task.getMeta().set(dlMCMetaConfig.flatMap(DownloadMCMeta::getOutput));
+            task.getOutput().set(project.getLayout().getBuildDirectory().dir("natives"));
         });
+
+        downloadAssets.configure(task -> task.getMeta().set(dlMCMetaConfig.flatMap(DownloadMCMeta::getOutput)));
+
         applyPatches.configure(task -> {
-            task.setOutput(project.file("build/" + task.getName() + "/output.zip"));
-            task.setRejects(project.file("build/" + task.getName() + "/rejects.zip"));
-            task.setPatches(extension.patches);
-            task.setPatchMode(PatchMode.ACCESS);
+            final Provider<Directory> workDir = project.getLayout().getBuildDirectory().dir(task.getName());
+            task.getOutput().set(workDir.map(s -> s.file("output.zip")));
+            task.getRejects().set(workDir.map(s -> s.file("rejects.zip").getAsFile()));
```

Comment:
If you look to within the `if` condition, you can see that the rejects 'output' is set to a directory, while here it is set to a file. Because of this duality of meaning, I could not use either `RegularFile` and `Directory`, so at that point I moved it to be a `Property<File>` instead. Though in hindsight it may be possible to move it to a `Property<FileSystemLocation<?>>`, but that would need more testing if Gradle is capable of handling that (it is the superinterface of `RegularFile` and `Directory`).

Reply:
`FileSystemLocationProperty` exists, but it fails when trying to use it as `FileSystemLocationProperty<FileSystemLocation>` because `FileSystemLocation` is an interface, and it cannot make a decorated property for it. Seems fine as-is for now.